homeassistant/components/edl21/sensor.py

```diff
@@ -107,16 +131,41 @@ def event(self, message_body) -> None:
                     self._OBIS_BLACKLIST.add(obis)
 
         if new_entities:
-            self._async_add_entities(new_entities, update_before_add=True)
+            self._hass.loop.create_task(self.add_entities(new_entities))
+
+    async def add_entities(self, new_entities) -> None:
+        """Migrate old unique IDs, then add entities to hass."""
+        registry = await async_get_registry(self._hass)
+
+        for entity in new_entities:
+            old_entity_id = registry.async_get_entity_id(
+                "sensor", DOMAIN, entity.old_unique_id
+            )
+            if old_entity_id is not None:
+                _LOGGER.debug(
+                    "Migrating unique_id from [%s] to [%s]",
+                    entity.old_unique_id,
+                    entity.unique_id,
+                )
+                if registry.async_get_entity_id("sensor", DOMAIN, entity.unique_id):
+                    registry.async_remove(old_entity_id)
+                else:
+                    registry.async_update_entity(
+                        old_entity_id, new_unique_id=entity.unique_id
+                    )
+
+        self._async_add_entities(new_entities, update_before_add=True)
 
 
 class EDL21Entity(Entity):
     """Entity reading values from EDL21 telegram."""
 
-    def __init__(self, obis, name, telegram):
+    def __init__(self, electricity_id, obis, name, telegram):
         """Initialize an EDL21Entity."""
+        self._electricity_id = electricity_id
         self._obis = obis
         self._name = name
+        self._unique_id = f"{DOMAIN}_{electricity_id}_{obis}"
```

Comment:
```suggestion
        self._unique_id = f"{electricity_id}_{obis}"
```

You can drop DOMAIN here as entity unique ids are unique to the domain.